proxy/proxy.ts

```diff
@@ -13,16 +14,178 @@ class BadRequestError extends Error {
   }
 }
 
-export function createProxyServer(config: {
+interface RateLimitInfo {
+  count: number
+  timestamp: number
+}
+
+interface ProxyConfig {
   allowLocalhost?: boolean
   allowsFrom: RegExp[]
   maxSize: number
   timeout: number
-}): Server {
-  return createServer(async (req, res) => {
+}
+
+const RATE_LIMIT = {
+  PER_DOMAIN: {
+    LIMIT: 500,
+    DURATION: 60 * 1000
+  },
+  GLOBAL: {
+    LIMIT: 5000,
+    DURATION: 60 * 1000
+  }
+}
+
+let rateLimitMap: Map<string, RateLimitInfo> = new Map()
+let requestQueue: Map<string, Promise<void>> = new Map()
+
+function isRateLimited(ip: string, domain: string): boolean {
+  let now = performance.now()
+
+  let domainKey = `${ip}:${domain}`
+  let domainRateLimit = rateLimitMap.get(domainKey) || {
+    count: 0,
+    timestamp: now
+  }
+
+  if (now - domainRateLimit.timestamp > RATE_LIMIT.PER_DOMAIN.DURATION) {
+    domainRateLimit.count = 0
+    domainRateLimit.timestamp = now
+  }
+
+  if (domainRateLimit.count >= RATE_LIMIT.PER_DOMAIN.LIMIT) {
+    return true
+  }
+
+  let globalKey = ip
+  let globalRateLimit = rateLimitMap.get(globalKey) || {
```

Comment:
The logic is same between `globalKey` and `domainKey`.

Maybe we should have function like:

```
function isRateLimited(key: string, store: Map<string, RateLimitInfo>, limit: { LIMIT: number, DURATION: number })
```